fix(mssql): support top syntax for limiting queries

sujiplr opened the pull request:
MSSQL is not supporting LIMIT syntax in SQLs. For limiting the rows, MSSQL having a different keyword TOP. Added fixes for handling the TOP and LIMIT clauses based on the database engines. Made this fix as generalised one for handling different database engines.

<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->

### SUMMARY
This fix will help to start supporting TOP clause in SQLs. And also along with CTE queries.

### BEFORE SCREENSHOTS

![Pre-error-with-cte](https://user-images.githubusercontent.com/31705464/154684448-9ea73d92-ec23-4cd0-9266-5abf734c20ad.gif)

### AFTER SCREENSHOTS

![Post-fix-working with top and cte](https://user-images.githubusercontent.com/31705464/154684554-47884774-cbc3-45b0-b310-f0889a254dfb.gif)

### TESTING INSTRUCTIONS
You can test the SQLs as per the above screenshots. 

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [x] Has associated issue: closes #8074
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

sujiplr committed:
SQL-TOP Fix For Database Engines

MSSQL is not supporting LIMIT syntax in SQLs. For limiting the rows, MSSQL having a different keyword TOP. Added fixes for handling the TOP and LIMIT clauses based on the database engines.

sujiplr committed:
Teradata code for top clause handling removed from teradata.py

Teradata code for top clause handling removed from teradata.py file, since we added generic section in base engine for the same.

villebro commented on superset/db_engine_specs/base.py:
```diff
@@ -649,6 +653,57 @@ def apply_limit_to_sql(
+        """
+        sql_statement = sqlparse.format(sql, strip_comments=True)
+        query_limit: Optional[int] = sql_parse._extract_top_from_query(sql_statement)
+        td_sel_keywords = {"SELECT","SEL"}
+        td_top_keywords = {"TOP","SAMPLE"}
```
`SEL` and `SAMPLE` seem to be specific to Teradata. We should probably move these into `BaseEngineSpec` like this:

```python
class BaseEngineSpec:
    ...
    # all keywords that should be treated as SELECTs
    select_keywords = {"SELECT"}
    # For engines that use TOP instead of LIMIT keyword, 
    list all keywords that should be treated similar top TOP keywords
    top_keywords = {"TOP"}
```

Then for Teradata these could be set to
```python
    select_keywords = {"SELECT","SEL"}
    top_keywords = {"TOP","SAMPLE"}
```

and for MSSQL it would probably be ok to go with the defaults (I assume `SEL` and `SAMPLE` aren't valid for MSSQL).

sujiplr replied:
Made this change

villebro commented on superset/sql_parse.py:
```diff
@@ -77,6 +77,29 @@ def _extract_limit_from_query(statement: TokenList) -> Optional[int]:
                 return int(token.value)
     return None
 
+def _extract_top_from_query(statement: TokenList) -> Optional[int]:
+    """
+    Extract top clause value from SQL statement.
+
+    :param statement: SQL statement
+    :return: top value extracted from query, None if no top value present in statement
+    """
+
+    td_top_keyword = {"TOP","SAMPLE"}
+    str_statement = str(statement)
+    str_statement = str_statement.replace("\n", " ").replace("\r", "")
+    token = str_statement.rstrip().split(" ")
+    token = [part for part in token if part]
+    top = None
+    for i, _ in enumerate(token):
+        if token[i].upper() in td_top_keyword and len(token) - 1 > i:
+            try:
+                top = int(token[i + 1])
+            except ValueError:
+                top = None
+            break
+    return top
```
Here we'd need to pass in the `top_keywords` from the db engine spec to make it possible to customize supported keywords per engine. Something like this:
```suggestion
def _extract_top_from_query(statement: TokenList, top_keywords: Set[str]) -> Optional[int]:
    """
    Extract top clause value from SQL statement.

    :param statement: SQL statement
    :return: top value extracted from query, None if no top value present in statement
    """

    str_statement = str(statement)
    str_statement = str_statement.replace("\n", " ").replace("\r", "")
    token = str_statement.rstrip().split(" ")
    token = [part for part in token if part]
    top = None
    for i, _ in enumerate(token):
        if token[i].upper() in top_keywords and len(token) - 1 > i:
            try:
                top = int(token[i + 1])
            except ValueError:
                top = None
            break
    return top
```

villebro reviewed (commented):
First pass comments. Let's also ping the original authors of the Teradata TOP functionality when this is ready 👍

sujiplr committed:
Changes to handle CTE along with TOP in complex SQLs

Added changes to handle TOP command in CTEs, for DB Engines which are not supporting inline CTEs.

sujiplr committed:
Test cases for TOP unit testing in MSSQL

Added multiple unit test cases for MSSQL top command handling and also along with CTEs

sujiplr committed:
Corrected the select_keywords name key in basengine

Corrected the select_keywords name key in basengine

villebro commented on superset/db_engine_specs/base.py:
```diff
@@ -638,8 +647,8 @@ def apply_limit_to_sql(
             sql = sql.strip("\t\n ;")
             qry = (
                 select("*")
-                .select_from(TextAsFrom(text(sql), ["*"]).alias("inner_qry"))
-                .limit(limit)
+                    .select_from(TextAsFrom(text(sql), ["*"]).alias("inner_qry"))
+                    .limit(limit)
```
I expect these changes to cause linting errors

sujiplr replied:
Corrected this

villebro commented on superset/db_engine_specs/base.py:
```diff
@@ -697,6 +770,7 @@ def get_cte_query(cls, sql: str) -> Optional[str]:
 
         return None
 
+
```
nit: This will also cause a linting error
```suggestion
```

sujiplr replied:
Corrected.

villebro commented on superset/db_engine_specs/mssql.py:
```diff
@@ -48,6 +48,9 @@ class MssqlEngineSpec(BaseEngineSpec):
     max_column_name_length = 128
     allows_cte_in_subquery = False
+    allow_limit_clause = False
+    sel_keywords = {"SELECT"}
+    top_keywords = {"TOP"}
```
wasn't this called `select_keywords` in `BaseEngineSpec`? Also, no need to override if the property is unchanged in the extending class.

sujiplr replied:
Corrected this.

villebro reviewed (commented):
Second pass comments

sujiplr commented on superset/db_engine_specs/base.py:
```diff
@@ -300,6 +299,16 @@ class BaseEngineSpec:  # pylint: disable=too-many-public-methods
+    select_keywords = {"SELECT"}
+    # This list will give the default list of keywords for data limit statements
+    # to consider in the limit/top SQl parsing
+    top_keywords = {"LIMIT"}
```
corrected this

sujiplr commented on superset/db_engine_specs/base.py:
```diff
@@ -649,6 +658,70 @@ def apply_limit_to_sql(
+        sql = " ".join(new_tokens)
+        return cte + sql
+
+    def top_not_in_sql(sql: str, top_words: Set[str]) -> bool:
+        for top_word in top_words:
+            if top_word.upper() in sql.upper():
+                return False
+        return True
```
Corrected this

sujiplr commented on superset/db_engine_specs/base.py:
```diff
@@ -649,6 +658,70 @@ def apply_limit_to_sql(
+
+        tokens = str_statement.rstrip().split(" ")
+        tokens = [token for token in tokens if token]
+        if cls.top_not_in_sql(str_statement, td_top_keywords):
```
Corrected this

sujiplr commented on superset/db_engine_specs/base.py:
```diff
@@ -649,6 +658,70 @@ def apply_limit_to_sql(
+        cte = None
+        sql_remainder = None
+        sql_statement = sqlparse.format(sql, strip_comments=True)
+        query_limit: Optional[int] = sql_parse._extract_top_from_query(sql_statement,
+                                                                       td_top_keywords)
```
Corrected this

sujiplr commented on superset/db_engine_specs/base.py:
```diff
@@ -649,6 +658,70 @@ def apply_limit_to_sql(
+        :return: SQL query with top clause
+        """
+
+        td_sel_keywords = cls.select_keywords
+        td_top_keywords = cls.top_keywords
```
Corrected this, instead of assignment referenced as class property value.

sujiplr committed:
Changes based on as per review.

made the required corrections based on code review to keep good code readability and code cleanliness.

villebro reviewed (commented):
Some last nit comments, but I think this is pretty much ready for a final test pass + a few added unit tests to the MSSQL test suite. Also, I suggest checking failing workflows on CI, as there appears to be simple linting issues that need to be addressed.

sujiplr commented on superset/db_engine_specs/base.py:
```diff
@@ -80,7 +80,6 @@
 
 logger = logging.getLogger()
 
-
```
Corrected.

sujiplr commented on superset/db_engine_specs/base.py:
```diff
@@ -300,6 +299,16 @@ class BaseEngineSpec:  # pylint: disable=too-many-public-methods
+    # If True, then the database engine is allowed for LIMIT clause
+    # If False, then the database engine is allowed for TOP clause
+    allow_limit_clause = True
+    # This list will give keywords for select statements
+    # to consider for the engines with TOP SQl parsing
+    select_keywords: Set[str] = None
+    # This list will give the keywords for data limit statements
+    # to consider for the engines with TOP SQl parsing
+    top_keywords: Set[str] = None
```
Corrected

sujiplr commented on superset/db_engine_specs/base.py:
```diff
@@ -649,6 +658,69 @@ def apply_limit_to_sql(
+        else:
+            final_limit = _limit
+        if not cls.allows_cte_in_subquery:
+            cte, sql_remainder = sql_parse.get_cte_reminder_query(sql_statement)
```
Corrected.

sujiplr commented on superset/sql_parse.py:
```diff
@@ -78,6 +78,56 @@ def _extract_limit_from_query(statement: TokenList) -> Optional[int]:
+    return top
+
+
+def get_cte_reminder_query(sql: str) -> Optional[str]:
```
Corrected

sujiplr commented on superset/sql_lab.py:
```diff
@@ -287,15 +287,27 @@ def execute_sql_statement(  # pylint: disable=too-many-arguments,too-many-locals
     return SupersetResultSet(data, cursor_description, db_engine_spec)
 
 
+def apply_cte(
+    database: Database, sql: str
+) -> str:
+    # Use the Basengine function to parse the SQL for CTE
+    sql = database.db_engine_spec.get_cte_query(sql)
+    return sql
+
+
```
removed this function.

sujiplr commented on superset/sql_parse.py:
```diff
@@ -30,7 +30,8 @@
     TokenList,
 )
-from sqlparse.tokens import DDL, DML, Keyword, Name, Punctuation, String, Whitespace
+from sqlparse.tokens import DDL, DML, Keyword, Name, Punctuation, String, Whitespace, \
+    CTE
```
Corrected.

sujiplr committed:
Review changes to correct lint and typo issues

Made the changes according to the review comments.

villebro committed:
Merge branch 'master' into SQL-TOP-FIX

villebro committed:
fix linting errors

villebro committed:
fix teradata tests

villebro committed:
add coverage

villebro committed:
lint

villebro commented on superset/sql_lab.py:
```diff
@@ -292,10 +292,14 @@ def apply_limit_if_exists(
+        # Engine support it will choose top or limit parse
         # Later, the extra row will be dropped before sending
         # the results back to the user.
-        sql = database.apply_limit_to_sql(sql, increased_limit, force=True)
+        if database.db_engine_spec.allow_limit_clause:
+            sql = database.apply_limit_to_sql(sql, increased_limit, force=True)
+        else:
+            sql = database.apply_top_to_sql(sql, increased_limit)
```
nit: this check should probably be done in `Database. apply_top_to_sql` so callers don't need to be aware of LIMIT/TOP.

sujiplr replied:
done

villebro commented on superset/sql_parse.py:
```diff
@@ -41,7 +50,6 @@
 CTE_PREFIX = "CTE__"
 logger = logging.getLogger(__name__)
 
-
```
nit: maybe add this back to avoid making unrelated changes

sujiplr replied:
Done.

villebro replied:
removed line seems to still be here

villebro reviewed (commented):
Two minor nits

sujiplr committed:
Code cleanliness

Moved the top/limit flag check from sql_lab to core.

villebro commented on superset/models/core.py:
```diff
@@ -488,7 +488,13 @@ def select_star(  # pylint: disable=too-many-arguments
+        else:
+            return self.db_engine_spec.apply_top_to_sql(sql, limit)
+
+    def apply_top_to_sql(self, sql: str, limit: int = 1000, force: bool = False) -> str:
```
I don't think this is needed anymore

sujiplr replied:
Done

villebro commented on superset/sql_lab.py:
```diff
@@ -290,9 +290,12 @@ def execute_sql_statement(  # pylint: disable=too-many-arguments,too-many-locals
     database: Database, increased_limit: Optional[int], query: Query, sql: str
 ) -> str:
+    print("apply_limit_if_exists")
+    print(database.db_engine_spec.allow_limit_clause)
```
please remove these

sujiplr replied:
Done

villebro commented on superset/db_engine_specs/base.py:
```diff
@@ -80,7 +80,6 @@
 
 logger = logging.getLogger()
 
-
```
removed newline seems to still be here

sujiplr replied:
done

sujiplr committed:
Changed for code cleanliness

Changes for keeping code cleanliness

sujiplr committed:
Corrected lint issue

Corrected lint issue.

villebro commented on superset/db_engine_specs/mssql.py:
```diff
@@ -48,6 +48,9 @@ class MssqlEngineSpec(BaseEngineSpec):
     max_column_name_length = 128
     allows_cte_in_subquery = False
+    allow_limit_clause = False
+    select_keywords = {"SELECT"}
+    top_keywords = {"TOP"}
```
these can be removed (they're identical to the defaults in `BaseEngineSpec`)

villebro reviewed (commented):
Last nit, after this I believe this is good to go

sujiplr committed:
Code cleanliness

Code cleanliness

villebro reviewed (approved):
LGTM! Thanks for all the hard work here; this is a major step forward in ensuring Superset works properly on MSSQL, Synapse and other Azure db products! ❤️

sfirke commented:
I am noodling on https://github.com/apache/superset/issues/21027#issuecomment-1516398570 and found this PR via blame view on the relevant code.  @sujiplr do you have thoughts on that issue - modifying the codebase to support use of `DISTINCT` in SQL Lab for SQL Server?  The current code injects `TOP 1000` in the wrong place for MSSQL so it's `SELECT TOP 1000 DISTINCT` when it needs to be `SELECT DISTINCT TOP 1000`.
